zwave_js_server/model/endpoint.py

```diff
@@ -4,27 +4,79 @@
 https://zwave-js.github.io/node-zwave-js/#/api/endpoint?id=endpoint-properties
 """
 
-from typing import Optional, TypedDict
+from typing import Dict, List, TYPE_CHECKING, Optional, TypedDict, Union
 
+from .command_class import CommandClassInfoDataType
 from .device_class import DeviceClass, DeviceClassDataType
+from .device_config import DeviceConfigDataType
+from ..event import EventBase
+from .value import ConfigurationValue, Value, ValueDataType
+
+if TYPE_CHECKING:
+    from ..client import Client
 
 
 class EndpointDataType(TypedDict, total=False):
     """Represent an endpoint data dict type."""
 
+    # Endpoint data
     nodeId: int  # required
     index: int  # required
     deviceClass: DeviceClassDataType  # required
     installerIcon: int
     userIcon: int
 
+    # Node data
```

Comment:
It looks weird to join the node data here since that data isn't used by the Endpoint class. I think we should still have two different data types and only pass the endpoint data to the Endpoint class.

Reply:
the underlying type for data would need to be `Union[EndpointDataType, NodeDataType]` in order to define separate datatypes but also have `Node` be a subclass of `Endpoint`. Are we ok with that?

Reply:
Why? To the Node we'll pass an instance of NodeDataType and to the Endpoint we'll pass a different instance of EndpointDataType.

Reply:
I was getting mypy errors when I tried this. I had to do the Union in order to get it to stop. Without the Union I get the following:
```
zwave_js_server/model/node.py:139: error: TypedDict "EndpointDataType" has no key 'status'
zwave_js_server/model/node.py:144: error: Returning Any from function declared to return "Optional[bool]"
zwave_js_server/model/node.py:144: error: TypedDict "EndpointDataType" has no key 'ready'
zwave_js_server/model/node.py:149: error: Returning Any from function declared to return "Optional[bool]"
zwave_js_server/model/node.py:149: error: TypedDict "EndpointDataType" has no key 'isListening'
zwave_js_server/model/node.py:154: error: Returning Any from function declared to return "Union[bool, str, None]"
zwave_js_server/model/node.py:154: error: TypedDict "EndpointDataType" has no key 'isFrequentListening'
zwave_js_server/model/node.py:159: error: Returning Any from function declared to return "Optional[bool]"
zwave_js_server/model/node.py:159: error: TypedDict "EndpointDataType" has no key 'isRouting'
zwave_js_server/model/node.py:164: error: Returning Any from function declared to return "Optional[int]"
zwave_js_server/model/node.py:164: error: TypedDict "EndpointDataType" has no key 'maxDataRate'
zwave_js_server/model/node.py:169: error: Returning Any from function declared to return "List[int]"
zwave_js_server/model/node.py:169: error: TypedDict "EndpointDataType" has no key 'supportedDataRates'
zwave_js_server/model/node.py:174: error: Returning Any from function declared to return "Optional[bool]"
zwave_js_server/model/node.py:174: error: TypedDict "EndpointDataType" has no key 'isSecure'
zwave_js_server/model/node.py:179: error: Returning Any from function declared to return "Optional[int]"
zwave_js_server/model/node.py:179: error: TypedDict "EndpointDataType" has no key 'protocolVersion'
zwave_js_server/model/node.py:184: error: Returning Any from function declared to return "Optional[bool]"
zwave_js_server/model/node.py:184: error: TypedDict "EndpointDataType" has no key 'supportsBeaming'
zwave_js_server/model/node.py:189: error: Returning Any from function declared to return "Optional[bool]"
zwave_js_server/model/node.py:189: error: TypedDict "EndpointDataType" has no key 'supportsSecurity'
zwave_js_server/model/node.py:194: error: Returning Any from function declared to return "Optional[int]"
zwave_js_server/model/node.py:194: error: TypedDict "EndpointDataType" has no key 'manufacturerId'
zwave_js_server/model/node.py:199: error: Returning Any from function declared to return "Optional[int]"
zwave_js_server/model/node.py:199: error: TypedDict "EndpointDataType" has no key 'productId'
zwave_js_server/model/node.py:204: error: Returning Any from function declared to return "Optional[int]"
zwave_js_server/model/node.py:204: error: TypedDict "EndpointDataType" has no key 'productType'
zwave_js_server/model/node.py:209: error: Returning Any from function declared to return "Optional[str]"
zwave_js_server/model/node.py:209: error: TypedDict "EndpointDataType" has no key 'firmwareVersion'
zwave_js_server/model/node.py:214: error: Returning Any from function declared to return "Optional[int]"
zwave_js_server/model/node.py:214: error: TypedDict "EndpointDataType" has no key 'zwavePlusVersion'
zwave_js_server/model/node.py:219: error: Returning Any from function declared to return "Optional[int]"
zwave_js_server/model/node.py:219: error: TypedDict "EndpointDataType" has no key 'zwavePlusNodeType'
zwave_js_server/model/node.py:224: error: Returning Any from function declared to return "Optional[int]"
zwave_js_server/model/node.py:224: error: TypedDict "EndpointDataType" has no key 'zwavePlusRoleType'
zwave_js_server/model/node.py:229: error: Returning Any from function declared to return "Optional[str]"
zwave_js_server/model/node.py:229: error: TypedDict "EndpointDataType" has no key 'name'
zwave_js_server/model/node.py:234: error: Returning Any from function declared to return "Optional[str]"
zwave_js_server/model/node.py:234: error: TypedDict "EndpointDataType" has no key 'location'
zwave_js_server/model/node.py:244: error: Returning Any from function declared to return "Optional[str]"
zwave_js_server/model/node.py:244: error: TypedDict "EndpointDataType" has no key 'label'
zwave_js_server/model/node.py:249: error: Returning Any from function declared to return "Optional[str]"
zwave_js_server/model/node.py:249: error: TypedDict "EndpointDataType" has no key 'deviceDatabaseUrl'
zwave_js_server/model/node.py:254: error: Returning Any from function declared to return "Optional[bool]"
zwave_js_server/model/node.py:254: error: TypedDict "EndpointDataType" has no key 'endpointCountIsDynamic'
zwave_js_server/model/node.py:259: error: Returning Any from function declared to return "Optional[bool]"
zwave_js_server/model/node.py:259: error: TypedDict "EndpointDataType" has no key 'endpointsHaveIdenticalCapabilities'
zwave_js_server/model/node.py:264: error: Returning Any from function declared to return "Optional[int]"
zwave_js_server/model/node.py:264: error: TypedDict "EndpointDataType" has no key 'individualEndpointCount'
zwave_js_server/model/node.py:269: error: Returning Any from function declared to return "Optional[int]"
zwave_js_server/model/node.py:269: error: TypedDict "EndpointDataType" has no key 'aggregatedEndpointCount'
zwave_js_server/model/node.py:274: error: Returning Any from function declared to return "Optional[int]"
zwave_js_server/model/node.py:274: error: TypedDict "EndpointDataType" has no key 'interviewAttempts'
zwave_js_server/model/node.py:279: error: Returning Any from function declared to return "Union[int, str, None]"
zwave_js_server/model/node.py:279: error: TypedDict "EndpointDataType" has no key 'interviewStage'
zwave_js_server/model/node.py:289: error: TypedDict "EndpointDataType" has no key 'commandClasses'
zwave_js_server/model/node.py:443: error: TypedDict "EndpointDataType" has no key 'status'
zwave_js_server/model/node.py:448: error: TypedDict "EndpointDataType" has no key 'status'
zwave_js_server/model/node.py:453: error: TypedDict "EndpointDataType" has no key 'status'
zwave_js_server/model/node.py:458: error: TypedDict "EndpointDataType" has no key 'status'
zwave_js_server/model/node.py:463: error: TypedDict "EndpointDataType" has no key 'ready'
zwave_js_server/model/node.py:464: error: TypedDict "EndpointDataType" has no key 'interviewStage'
zwave_js_server/model/node.py:468: error: TypedDict "EndpointDataType" has no key 'interviewStage'
zwave_js_server/model/node.py:473: error: TypedDict "EndpointDataType" has no key 'interviewStage'
zwave_js_server/model/node.py:478: error: TypedDict "EndpointDataType" has no key 'ready'
```

Reply:
Right. It's because we use the same instance attribute for the data parameter for the parent and the child.

I don't know. This inheritance model smells funny to me.

Reply:
OK well there's nothing preventing us from not merging this and moving forward with the other PRs. This just seemed like the right thing to do since that's how zwave-js is modeled

Reply:
Setting the type on `self.data` in `Node` fixed the issue btw